internal/expr/prepare.go

```diff
@@ -125,81 +141,99 @@ func prepareOutput(ti typeNameToInfo, p *outputPart) ([]fullName, []typeMember,
 			if info, err = fetchInfo(t.prefix); err != nil {
-				return nil, nil, err
+				return err
 			}
 			// Generate asterisk columns.
 			if t.name == "*" {
 				switch info := info.(type) {
 				case *mapInfo:
-					return nil, nil, fmt.Errorf(`&%s.* cannot be used for maps when no column names are specified`, info.typ().Name())
+					return fmt.Errorf(`&%s.* cannot be used for maps when no column names are specified`, info.typ().Name())
 				case *structInfo:
 					if len(info.tags) == 0 {
-						return nil, nil, fmt.Errorf("type %q in %q does not have any db tags", info.typ().Name(), p.raw)
+						return fmt.Errorf("type %q in %q does not have any db tags", info.typ().Name(), p.raw)
 					}
 					for _, tag := range info.tags {
-						outCols = append(outCols, fullName{pref, tag})
-						typeMembers = append(typeMembers, info.tagToField[tag])
+						err := addSQL(fullName{prefix: pref, name: tag}.String(), info.tagToField[tag])
+						if err != nil {
+							return err
+						}
 					}
 				}
 			} else {
 				// Generate explicit columns.
-				if err = addColumns(info, t.name, fullName{pref, t.name}); err != nil {
-					return nil, nil, err
+				if err = addColumn(info, t.name, fullName{prefix: pref, name: t.name}); err != nil {
+					return err
 				}
 			}
 		}
-		return outCols, typeMembers, nil
+		return nil
 	} else if numColumns > 1 && starColumns > 0 {
-		return nil, nil, fmt.Errorf("invalid asterisk in output expression columns: %s", p.raw)
+		return fmt.Errorf("invalid asterisk in output expression columns: %s", p.raw)
 	}
 
 	// Case 2: Explicit columns, single asterisk type e.g. "(col1, t.col2) AS &P.*".
 	if starTypes == 1 && numTypes == 1 {
 		if info, err = fetchInfo(p.targetTypes[0].prefix); err != nil {
-			return nil, nil, err
+			return err
 		}
 		for _, c := range p.sourceColumns {
-			if err = addColumns(info, c.name, c); err != nil {
-				return nil, nil, err
+			switch c := c.(type) {
+			case funcExpr:
+				return fmt.Errorf(`cannot use function %q with asterisk output expression: %q`, c.raw, p.raw)
+			case fullName:
+				if err = addColumn(info, c.name, c); err != nil {
+					return err
+				}
 			}
 		}
-		return outCols, typeMembers, nil
+		return nil
 	} else if starTypes > 0 && numTypes > 1 {
-		return nil, nil, fmt.Errorf("invalid asterisk in output expression types: %s", p.raw)
+		return fmt.Errorf("invalid asterisk in output expression types: %s", p.raw)
 	}
 
 	// Case 3: Explicit columns and types e.g. "(col1, col2) AS (&P.name, &P.id)".
 	if numColumns == numTypes {
 		for i, c := range p.sourceColumns {
 			t := p.targetTypes[i]
 			if info, err = fetchInfo(t.prefix); err != nil {
-				return nil, nil, err
+				return err
 			}
 
-			if err = addColumns(info, t.name, c); err != nil {
-				return nil, nil, err
+			switch c := c.(type) {
+			case funcExpr:
+				if err = pe.prepareParts(ti, c.pe.queryParts); err != nil {
+					return err
+				}
+				err = addColumn(info, t.name, fullName{})
```

Comment:
Why are we passing here an empty column name? Is the intention that the `addColumn` function prints only the "AS ..." part in the line below?
```
pe.sql += column + " AS " + markerName(len(pe.outputs)) + ", "
```